### PR TITLE
HPCC-25018 Add support for dfu export and import to cloud

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Pushed to run smoke test.  Possibly ready for review, but I plan to review myself and test some more before declaring it ready.

**Comment by ghalliday** [comment]
@jakesmith any idea why the thor test failed?

**Comment by jakesmith** [comment]
> any idea why the thor test failed?

@ghalliday - hm not sure. It was the 1st query in that batch (action1a) that failed, but it wasn't the query itself. 1 of the 2 slave was aborted by something:
{noformat}
00000016 USR 2021-01-22 17:23:01.780 10233 10233 "Slave log 1 contains slave: 1"
00000017 USR 2021-01-22 17:23:01.782 10233 10286 "================================================"
00000018 USR 2021-01-22 17:23:01.782 10233 10286 "Program:   10.1.0.4:/home/runner/work/HPCC-Platform/HPCC-Platform/install/opt/HPCCSystems/bin/thorslave_lcr"
00000019 USR 2021-01-22 17:23:01.782 10233 10286 "Signal:    6 Aborted"
0000001A USR 2021-01-22 17:23:01.782 10233 10286 "Fault IP:  00007FA8F39B918B"
0000001B USR 2021-01-22 17:23:01.782 10233 10286 "Accessing: 000003E9000027F9"
0000001C PRG 2021-01-22 17:23:01.782 10233 10286 "Backtrace:"
0000001D PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb) [0x7fa8f39b918b]"
0000001E PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libc.so.6(abort+0x12b) [0x7fa8f3998859]"
0000001F PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libstdc++.so.6(+0x9e951) [0x7fa8f3c43951]"
00000020 PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa47c) [0x7fa8f3c4f47c]"
00000021 PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libstdc++.so.6(+0xaa4e7) [0x7fa8f3c4f4e7]"
00000022 PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libstdc++.so.6(+0xab245) [0x7fa8f3c50245]"
00000023 PRG 2021-01-22 17:23:01.783 10233 10286 "  ./thorslave_mythor(+0x26ebd) [0x55969eccfebd]"
00000024 PRG 2021-01-22 17:23:01.783 10233 10286 "  ./thorslave_mythor(_ZThn16_N9CThreaded3runEv+0x14) [0x55969ecbb394]"
00000025 PRG 2021-01-22 17:23:01.783 10233 10286 "  /home/runner/work/HPCC-Platform/HPCC-Platform/install/opt/HPCCSystems/lib/libjlib.so(_ZN6Thread5beginEv+0x32) [0x7fa8f3fbcfa2]"
00000026 PRG 2021-01-22 17:23:01.783 10233 10286 "  /home/runner/work/HPCC-Platform/HPCC-Platform/install/opt/HPCCSystems/lib/libjlib.so(_ZN6Thread11_threadmainEPv+0x26) [0x7fa8f3fbc016]"
00000027 PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libpthread.so.0(+0x9609) [0x7fa8f3b70609]"
00000028 PRG 2021-01-22 17:23:01.783 10233 10286 "  /lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7fa8f3a95293]"
00000029 USR 2021-01-22 17:23:01.783 10233 10286 "Registers:"
0000002A USR 2021-01-22 17:23:01.783 10233 10286 "EAX:0000000000000000  EBX:00007FA8E8EE3700  ECX:00007FA8F39B918B  EDX:0000000000000000  ESI:00007FA8E8EE21F0  EDI:0000000000000002"
0000002B USR 2021-01-22 17:23:01.783 10233 10286 "R8 :0000000000000000  R9 :00007FA8E8EE21F0  R10:0000000000000008  R11:0000000000000246"
0000002C USR 2021-01-22 17:23:01.783 10233 10286 "R12:00007FA8E8EE2510  R13:00007FFCB645359F  R14:00007FA8E8EE24F0  R15:00007FA8E8EE24D0"
0000002D USR 2021-01-22 17:23:01.783 10233 10286 "CS:EIP:0033:00007FA8F39B918B"
0000002E USR 2021-01-22 17:23:01.783 10233 10286 "   ESP:00007FA8E8EE21F0  EBP:0000000000000000"
0000002F USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE21F0]: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000"
00000030 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE2210]: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000"
00000031 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE2230]: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000"
00000032 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE2250]: 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 0000000000000000 7FFFFFFF00000000"
00000033 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE2270]: FFFFFFFE7FFFFFFF FFFFFFFFFFFFFFFE FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF"
00000034 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE2290]: FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF"
00000035 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE22B0]: FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF"
00000036 USR 2021-01-22 17:23:01.783 10233 10286 "Stack[00007FA8E8EE22D0]: FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF FFFFFFFFFFFFFFFF 00000000FFFFFFFF"
{noformat}

.. almost as soon as it started.
The other slave and the master have no logging to indicate they knew anything about it, until the master noticed that slave had disappeared a little later (@ 17:23:20).

**Comment by richardkchapman** [comment]
Signal 6 is usually what you get from an assert failing (in a debug build). Possibly also from a sanitize check failure if those are enabled?

**Comment by ghalliday** [comment]
@richardkchapman we should consider merging this to 7.12.x instead.  Then we could say that to migrate your data your bare metal system needs to be on the latest 7.12.x build.

File: dali/dfu/dfuutil.cpp [hunk 1]
```diff
@@ -1062,6 +1062,31 @@ class CDFUhelper: implements IDFUhelper, public CInterface
         else if (0 == strcmp(nodeName, queryDfsXmlBranchName(DXB_File)))
         {
             // Logical file map
+            //Patch the cluster before the file is created
+            if (!isEmptyString(cluster))
+            {
+                //In containerized mode the directory is configured somewhere else...
+                //replace @directory with a relative directory calculated from the scope
```
Comment (ghalliday) [inline dali/dfu/dfuutil.cpp:1069]: comment is out of date - it is absolute...

**Comment by jakesmith** [comment]
>Signal 6 is usually what you get from an assert failing (in a debug build). Possibly also from a sanitize check failure if those are enabled?

It's a RelWithDebInfo build, and --sanitize isn't on.
I want to enable sanitize on smoketest builds, I cleared up a few problems when I used it last time, but I think there are still some left that would cause failures - I'll revisit.

File: dali/dfu/dfuutil.cpp [hunk 2]
```diff
@@ -1062,6 +1062,31 @@ class CDFUhelper: implements IDFUhelper, public CInterface
...
+                    if (!sep)
+                        break;
+                    addPathSepChar(location);
+                    location.appendLower(sep - temp, temp);
+                    temp = sep + 2;
+                }
+                t->setProp("@directory", location);
+
+                //Remove any clusters in the incoming definition, and replace it with a single entry
+                t->removeProp("Cluster");
```
Comment (jakesmith) [inline dali/dfu/dfuutil.cpp:1086]: to remove all "Cluster"'s you'd need : 
while (t->removeProp("Cluster"));

File: dali/dfuplus/dfuplus.cpp [hunk 3]
```diff
@@ -1427,9 +1425,12 @@ int CDfuPlusHelper::add()
         int len = buf.length();
         xmlbuf.setBuffer(len, buf.detach(), true);
 
+        const char* dstcluster = globals->queryProp("dstcluster");
```
Comment (jakesmith) [inline dali/dfuplus/dfuplus.cpp:1428]: should there be a check here that it's a non-empty string?

File: esp/services/ws_fs/ws_fsService.cpp [hunk 4]
```diff
@@ -2328,6 +2328,24 @@ void CFileSprayEx::getDropZoneInfoByIP(double clientVersion, const char* ip, con
     }
 }
 
+static StringBuffer & expandLogicalAsPhysical(StringBuffer & target, const char * name, const char * separator)
```
Comment (jakesmith) [inline esp/services/ws_fs/ws_fsService.cpp:2331]: there is a similar function in dafdesc (makeSinglePhysicalPartName),
but it comes with baggage, e.g. it uses queryBaseDirectory().

So a new cleaner method like this is fine, should it go somewhere common like dautils.hpp ?

**Review by jakesmith — COMMENTED** [review]
@ghalliday - a few comments, 1 minor issue

File: dali/dfu/dfuutil.cpp (hunk 2 above)
Comment (ghalliday) [reply dali/dfu/dfuutil.cpp:1086]: @jakesmith actually I think the original code is correct - the ptree code removes all items that match the property.

Comment (ghalliday) [reply dali/dfu/dfuutil.cpp:1086]:
```
    Owned<IPropertyTree> pt = createPTree("me");
    pt->addPropTree("b", createPTree());
    pt->addPropTree("b", createPTree());
    pt->addPropTree("c", createPTree());
    pt->addPropTree("d", createPTree());
    printXML(pt);
    pt->removeProp("b");
    printXML(pt);


```

Comment (jakesmith) [reply dali/dfu/dfuutil.cpp]: hmmm - that's a bug.
If you remove a property with a qualifier or with a wildcard, it removes 1st match.
I see from the impl. though if it's a simple id, it removes the entry from the children, which may be a list container.

It should remove the 1st. To remove all there should be a loop, as there are elsewhere.

I still recommend changing it to while(.. removeProp ..) even if it is spuriously removing all on the 1st attempt.

I'll open a JIRA to look at the semantic difference, but am concerned re. changing it.. what will it break.
Further discussion in that JIRA.

Comment (jakesmith) [reply dali/dfu/dfuutil.cpp]: https://track.hpccsystems.com/browse/HPCC-25281

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good

**Pull request merged**